### PR TITLE
Feature/rfs 39 new federation workflow

### PR DESCRIPTION
**Pull request by amendelzon**
Implementing new federation workflow. Basically, a few methods have been added to the bridge:

- createFederation(n) with n being the threshold of the new federation
- addFederatorPublicKey(key) with key being the BTC public key of the federator to add
- removeFederatorPublicKey(key) with key being the BTC public key of the federator to remove
- commitFederation()
- rollbackFederation()

Once a new federation is committed, the old federation is marked as retiring and the new is put in place. BTC release will be attempted always from the new federation, whereas lock will work for both until the retiring federation retires.

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java [hunk 1]
```diff
@@ -132,7 +116,23 @@ private BridgeSerializationUtils(){}
...
+    public static byte[] serializeUTXOList(List<UTXO> list) throws IOException {
+        int nutxos = list.size();
+
+        byte[][] bytes = new byte[nutxos][];
+        int n = 0;
+
+        for (UTXO utxo : list) {
+            ByteArrayOutputStream ostream = new ByteArrayOutputStream();
+            utxo.serializeToStream(ostream);
+            ostream.close();
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java]: if `serializeToStream` throws an exception, the *ostream* never gets closed. You can use the _try with resources_ syntax here like this:
```java
try(ByteArrayOutputStream ostream = new ByteArrayOutputStream()) {
    utxo.serializeToStream(ostream);
    bytes[n++] = RLP.encodeElement(ostream.toByteArray());
}
```

File: rskj-core/src/main/java/co/rsk/peg/PendingFederation.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        );
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%d of %d signatures pending federation (%s)", numberOfSignaturesRequired, publicKeys.size(), isComplete() ? "complete" : "incomplete");
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Comment (diega) [inline rskj-core/src/main/java/co/rsk/peg/PendingFederation.java]: Overriding only equals violates the [hashcode](http://java.sun.com/j2se/1.5.0/docs/api/java/lang/Object.html#hashCode%28%29) contract of the Object. As a general rule, both methods (equals and hashcode) must be overridden together

File: rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java (hunk 1 above)
Comment (donequis) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java]: java.io.ByteArrayOutputStream.close() is an empty method

From the impl
```Java
    public void close() throws IOException {
    }
```

File: rskj-core/src/main/java/co/rsk/peg/PendingFederation.java (hunk 2 above)
Comment (donequis) [reply rskj-core/src/main/java/co/rsk/peg/PendingFederation.java]: Where's this being used? Can't we remove the @Override?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java (hunk 1 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java]: What @donequis said. And plus, this method didn't actually change, only its name.

Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSerializationUtils.java]: That doesn't necessarily mean we don't actually need to call the `close()`, just saying its been there for a while.

File: rskj-core/src/main/java/co/rsk/peg/PendingFederation.java (hunk 2 above)
Comment (amendelzon) [reply rskj-core/src/main/java/co/rsk/peg/PendingFederation.java]: Not being used. What's the normal way to implement `hashCode`, @diega ?

**Comment by rskops** [comment]
SonarQube analysis reported 147 issues
* ![MAJOR][MAJOR] 94 major
* ![MINOR][MINOR] 24 minor
* ![INFO][INFO] 29 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L39](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L39): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L43](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L43): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L48](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L48): Move the "" string literal on the left side of this string comparison. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1132)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L59](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L59): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L60](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L60): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L61](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L61): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L62](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L62): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L63](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L63): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L64](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L64): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)
1. ![MAJOR][MAJOR] [GenNodeKeyId.java#L65](https://github.com/rsksmart/rskj/blob/e3b640d0bd0823dc3f3f30e2126b26adf86c2d61/rskj-core/src/main/java/co/rsk/GenNodeKeyId.java#L65): Replace this use of System.out or System.err by a logger. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS106)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**